### PR TITLE
feat: Add valkey-check-acl offline ACL file validator

### LINKED ISSUE #3734: Add `valkey-check-acl` — offline ACL configuration file validator
### Problem

There is no way to validate ACL configuration files offline before loading them into a running server. Invalid ACL rules are only discovered at runtime when `ACL LOAD` fails, which can cause service disruptions.

External systems (e.g., ElastiCache) maintain hand-rolled reimplementations of Valkey's ACL validation logic. Each new Valkey version requires updating these external validators to match new commands and categories. A canonical validation tool shipped with Valkey would eliminate this duplication.

### Proposed Solution

A new `valkey-check-acl` tool that validates ACL files offline, similar in spirit to `valkey-check-rdb` and `valkey-check-aof`. The tool reuses Valkey's actual ACL parsing code (`ACLStringSetUser`, `ACLSetSelector`, etc.) so validation results are guaranteed to match server behavior.

### Features

- **Input modes**: ACL file, valkey.conf (auto-detected), stdin
- **Version-gated validation** (`--version`): Reject syntax and commands not available in a target version (e.g., selectors require 7.0, `db=` requires 9.1)
- **Validation levels** (`--level`): syntax (structure only), semantic (default, validates command names), full (adds warnings for risky patterns)
- **Simplification** (`--simplify`): Output minimal canonical ACL rules (removes redundancies, normalizes keys/channels/dbs)
- **External commands** (`--commands-file`): Load module command definitions for validation
- **Machine-parseable output** (`--json`)
- **Strict by default**: Unknown commands are errors (use `--ignore-unknown-commands` to relax)
- **Duplicate user detection**: Matches Valkey's behavior
- **Distinct exit codes**: 0=OK, 1=syntax errors, 2=semantic errors, 3=warnings only

### Examples

```bash
# Basic validation
valkey-check-acl /etc/valkey/users.acl

# Validate against Redis 7.0 syntax
valkey-check-acl --version 7.0 acl.conf

# Simplify rules to canonical form
valkey-check-acl --simplify acl.conf

# Validate with module commands
valkey-check-acl --commands-file search-module.conf acl.conf

# Pipe from stdin
echo "user alice on >pass ~* +@all" | valkey-check-acl -
```

### Implementation

Built as a multi-call binary (symlink to `valkey-server`), following the established pattern. Includes Redis compatibility symlink (`redis-check-acl`).

### PR DESCRIPTION
**Pull request by yulazariy**
## Description

Add a new tool that validates ACL configuration files offline without starting a server. It reuses Valkey's actual ACL parsing code so validation results are guaranteed to match server behavior.

### Motivation

- No way to validate ACL files before `ACL LOAD` — errors only discovered at runtime
- External systems maintain reimplementations of ACL validation that drift with each release
- A canonical tool eliminates duplication and ensures correctness

### Features

| Feature | Flag |
|---------|------|
| Version-gated validation | `--version 7.0` |
| Validation levels | `--level syntax\|semantic\|full` |
| Rule simplification | `--simplify` |
| Module commands | `--commands-file <path>` |
| JSON output | `--json` |
| Strict unknown commands | `--ignore-unknown-commands` to relax |
| Fail fast | `--fail-fast` |

### Files Changed

| File | Change |
|------|--------|
| `src/valkey-check-acl.c` | New — tool implementation (1000 lines) |
| `tests/integration/check-acl.tcl` | New — 111 integration tests |
| `src/server.c` | Dispatch for `valkey-check-acl` / `redis-check-acl` |
| `src/server.h` | Declare `valkey_check_acl_main()`, `ACLFreeUser()` |
| `src/Makefile` | Build rules, install symlink |
| `src/CMakeLists.txt` | CMake symlink |
| `cmake/Modules/SourceFiles.cmake` | Add to server sources |
| `tests/support/set_executable_path.tcl` | Test binary path |

### Testing

```bash
./runtest --single tests/integration/check-acl.tcl
# 111 tests pass
```

Tests cover: all input modes, validation levels, version gating (every boundary), command version checking, --ignore-unknown-commands, --commands-file, --simplify, exit codes, edge cases (duplicates, large files, selectors, subcommands).

Fixes #3734

### CODE REVIEW

File: src/valkey-check-acl.c [hunk 1]
```diff
@@ -0,0 +1,1025 @@
...
+ * Error collection
+ * --------------------------------------------------------------------------- */
+
+typedef struct {
+    sds *entries;
+    int count;
+    int capacity;
+} checkAclErrors;
+
+static void checkAclErrorsInit(checkAclErrors *e) {
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: nit: (here and everywhere): the prefix is a bit misleading, it sounds like the beginning of an action.
I would rename this into something like: `aclCheckerXXXX`

Comment (eifrah-aws) [reply src/valkey-check-acl.c]: 1. I would run here `memset(..)` on the struct - this will cover any new members in the future
2. Maybe add a sanity check `serverAssert(e != NULL)`

File: src/valkey-check-acl.c [hunk 2]
```diff
@@ -0,0 +1,1025 @@
...
+    e->entries = NULL;
+    e->count = 0;
+    e->capacity = 0;
+}
+
+static void checkAclErrorsAdd(checkAclErrors *e, sds msg) {
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: Why not using `list`

File: src/valkey-check-acl.c [hunk 3]
```diff
@@ -0,0 +1,1025 @@
...
+
+/* ---------------------------------------------------------------------------
+ * Error collection
+ * --------------------------------------------------------------------------- */
+
+typedef struct {
+    sds *entries;
+    int count;
+    int capacity;
+} checkAclErrors;
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: `list` ?

File: src/valkey-check-acl.c [hunk 4]
```diff
@@ -0,0 +1,1025 @@
...
+/* Returns 1 if v >= o. */
+static inline int versionGE(const checkAclVersion *v, const checkAclVersion *o) {
+    if (v->major != o->major) return v->major > o->major;
+    if (v->minor != o->minor) return v->minor > o->minor;
+    return v->patch >= o->patch;
+}
+
+typedef struct {
+    int level;
+    int json;
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c:65]: I think we support `json`

File: src/valkey-check-acl.c [hunk 5]
```diff
@@ -0,0 +1,1025 @@
...
+    int verbose;
+    int quiet;
+    int fail_fast;
+    int ignore_unknown_commands;
+    int simplify;
+    const char *filename; /* NULL means stdin */
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c:71]: If this is implied, why do we need another flag?

File: src/valkey-check-acl.c [hunk 6]
```diff
@@ -0,0 +1,1025 @@
...
+
+/* Returns 1 if v >= o. */
+static inline int versionGE(const checkAclVersion *v, const checkAclVersion *o) {
+    if (v->major != o->major) return v->major > o->major;
+    if (v->minor != o->minor) return v->minor > o->minor;
+    return v->patch >= o->patch;
+}
+
+typedef struct {
+    int level;
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c:64]: nit: we can collect all the flags inside a single `flags` member with bitwise. Future wise, this is easier to maintain than having N variables

File: src/valkey-check-acl.c [hunk 7]
```diff
@@ -0,0 +1,1025 @@
...
+static void extCommandsAdd(checkAclExtCommands *ext, sds name, checkAclVersion since) {
+    if (ext->count == ext->capacity) {
+        ext->capacity = ext->capacity ? ext->capacity * 2 : 64;
+        ext->entries = zrealloc(ext->entries, sizeof(checkAclExtCommand) * ext->capacity);
+    }
+    ext->entries[ext->count].name = name;
+    ext->entries[ext->count].since = since;
+    ext->count++;
+}
+
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: `list`?

File: src/valkey-check-acl.c [hunk 8]
```diff
@@ -0,0 +1,1025 @@
...
+    return NULL;
+}
+
+static void extCommandsFree(checkAclExtCommands *ext) {
+    for (int i = 0; i < ext->count; i++) sdsfree(ext->entries[i].name);
+    zfree(ext->entries);
+}
+
+/* Load a commands file. Returns 0 on success, -1 on error. */
+static int loadCommandsFile(const char *path, checkAclExtCommands *ext) {
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: Its worth adding a comment about the expected file format.

File: src/valkey-check-acl.c [hunk 9]
```diff
@@ -0,0 +1,1025 @@
...
+
+/* ---------------------------------------------------------------------------
+ * Version-gated token validation
+ *
+ * Pre-checks ACL tokens against the requested version BEFORE passing them
+ * to ACLStringSetUser. Returns NULL if the token is valid for the version,
+ * or an error string (caller must sdsfree) if it's not.
+ * --------------------------------------------------------------------------- */
+
+static sds checkTokenVersion(const char *op, size_t oplen, const checkAclVersion *ver) {
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: nit: change the function signature:
- Return `1` or `0` on success / failure
- Return the error string as an input argument


Another option is to use the same method used below in the function `validateUserLine` -> pass the errors list and just append the error to the error collector

File: src/valkey-check-acl.c [hunk 10]
```diff
@@ -0,0 +1,1025 @@
...
+            char *close = strchr(open, ')');
+            if (!close) {
+                suffix = sdscat(suffix, open);
+                break;
+            }
+            /* Extract inner content and simplify it. */
+            sds inner = sdsnewlen(open + 1, close - open - 1);
+            /* Find command rules inside selector. */
+            char *inner_cmd = strstr(inner, "+@all");
+            if (!inner_cmd) inner_cmd = strstr(inner, "-@all");
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: nit: here and everywhere, I would use `{` / `}` even for a short (one-liner) if statement

File: src/valkey-check-acl.c [hunk 11]
```diff
@@ -0,0 +1,1025 @@
...
+    /* Track seen usernames for duplicate detection. */
+    int seen_cap = 0, seen_count = 0;
+    sds *seen_users = NULL;
+
+    for (int i = 0; i < totlines; i++) {
+        int linenum = i + 1;
+        lines[i] = sdstrim(lines[i], " \t\r\n");
+
+        /* Skip blank lines and comments. */
+        if (lines[i][0] == '\0' || lines[i][0] == '#') continue;
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: Do we support comment in the middle of the line?

e.g. 

```
a = b # Some comment
```

File: src/valkey-check-acl.c [hunk 12]
```diff
@@ -0,0 +1,1025 @@
...
+
+        /* In config mode, skip non-user lines. */
+        if (is_config && strncasecmp(lines[i], "user ", 5) != 0) continue;
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: what happens if we put a TAB here? (`user\t`)

File: src/valkey-check-acl.c [hunk 13]
```diff
@@ -0,0 +1,1025 @@
...
+            if (config->fail_fast) break;
+            continue;
+        }
+        if (argc == 0) {
+            sdsfreesplitres(argv, argc);
+            continue;
+        }
+
+        /* Duplicate user detection (matches Valkey behavior: reject duplicates). */
+        if (argc >= 2 && !strcasecmp(argv[0], "user")) {
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c:650]: use dict for seen users?

File: src/valkey-check-acl.c [hunk 14]
```diff
@@ -0,0 +1,1025 @@
...
+                                  sdscatprintf(sdsempty(), "%s:%d: Duplicate user '%s'. A user can only be defined once.",
+                                               display_name, linenum, argv[1]));
+                has_semantic_error = 1;
+                error_count++;
+                sdsfreesplitres(argv, argc);
+                if (config->fail_fast) break;
+                continue;
+            }
+            /* Track this username. */
+            if (seen_count == seen_cap) {
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: use `list` (or per my previous suggestion, use `dict`)

File: src/valkey-check-acl.c [hunk 15]
```diff
@@ -0,0 +1,1025 @@
...
+    }
+
+    for (int s = 0; s < seen_count; s++) sdsfree(seen_users[s]);
+    zfree(seen_users);
+    sdsfreesplitres(lines, totlines);
+
+    if (config->verbose && error_count == 0)
+        fprintf(stderr, "Validated %d user(s).\n", user_count);
+
+    /* Return: 0 = no errors, 1 = syntax errors, 2 = semantic errors */
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: move this comment to the top of the function

File: src/valkey-check-acl.c [hunk 16]
```diff
@@ -0,0 +1,1025 @@
...
+    if (error_count == 0) return 0;
+    return has_semantic_error ? 2 : 1;
+}
+
+/* ---------------------------------------------------------------------------
+ * Output formatting
+ * --------------------------------------------------------------------------- */
+
+static void printResultsText(checkAclErrors *errors, checkAclWarnings *warnings, checkAclConfig *config) {
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: This seems like a repeating code (we have similar code in valkey-cli).
Can't we add a single c json library? for example, there is https://github.com/DaveGamble/cJSON with MIT license. There are other projects, but this one is basically 2 files

File: src/valkey-check-acl.c [hunk 17]
```diff
@@ -0,0 +1,1025 @@
...
+    config->fail_fast = 0;
+    config->ignore_unknown_commands = 0;
+    config->simplify = 0;
+    config->num_commands_files = 0;
+    config->filename = NULL;
+    config->is_stdin = 0;
+    config->version_set = 0;
+    config->version.major = 0;
+    config->version.minor = 0;
+    config->version.patch = 0;
```
Comment (eifrah-aws) [inline src/valkey-check-acl.c]: For other structures, you add `Init` method - why not here?

**Review by eifrah-aws — CHANGES_REQUESTED** [review]

File: src/valkey-check-acl.c (hunk 11 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: No. 
Valkey's own ACL parser (`ACLLoadConfiguredUsers`, `ACLLoadFromFile`) does not support inline comments — `#` is only treated as a comment when it's the first non-whitespace character on a line. We match server behavior exactly.

File: src/valkey-check-acl.c (hunk 1 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: fix both

File: src/valkey-check-acl.c (hunk 2 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: fixed

File: src/valkey-check-acl.c (hunk 3 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: fixed

File: src/valkey-check-acl.c (hunk 5 above)
Comment (yulazariy) [reply src/valkey-check-acl.c:71]: removed

File: src/valkey-check-acl.c (hunk 17 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: fixed

File: src/valkey-check-acl.c (hunk 16 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: We can do it in a dedicated PR

File: src/valkey-check-acl.c (hunk 15 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: fixed

File: src/valkey-check-acl.c (hunk 14 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: I change to dict

File: src/valkey-check-acl.c (hunk 13 above)
Comment (yulazariy) [reply src/valkey-check-acl.c:650]: fixed

File: src/valkey-check-acl.c (hunk 7 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: fixed (to dict)

File: src/valkey-check-acl.c (hunk 8 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: fixed

File: src/valkey-check-acl.c (hunk 10 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: fixed

File: src/valkey-check-acl.c (hunk 12 above)
Comment (yulazariy) [reply src/valkey-check-acl.c]: fixed
